website/content/docs/custom-widgets.md

```diff
@@ -198,3 +195,276 @@ You can also return a promise from `isValid`. While the promise is pending, the
 ```
 
 **Note:** Do not create a promise inside `isValid` - `isValid` is called right before trying to persist. This means that even if a previous promise was already resolved, when the user hits 'save', `isValid` will be called again. If it returns a new promise, it will be immediately marked as "in error" until the new promise resolves.
+
+### Writing custom widgets as a separate package
+
+Widgets are inputs for the Netlify CMS editor interface. It's a React component that receives user input and outputs a serialized value. Those are the only rules - the component can be extremely simple, like text input, or extremely complicated, like a full-blown markdown editor. They can make calls to external services, and generally do anything that JavaScript can do.
+
+For writing custom widgets as a separate package you should follow these steps:
+
+1. Create a directory 
+
+   ```javascript
+   mkdir my-custom-widget
+   ```
+2. Navigate to the directory 
+
+   ```javascript
+   cd my-custom-widget
+   ```
+3. For setting up a new npm package run this command:
+
+   ```javascript
+   npm init
+   ```
+4. Create a `src` directory with the files `Control.js`, `Preview.js` and `index.js`
+
+`src/Control.js`
+
+```javascript
+ import PropTypes from 'prop-types';
+ import React from 'react';
+
+ export default class Control extends React.Component {
+   static propTypes = {
+     onChange: PropTypes.func.isRequired,
+     forID: PropTypes.string,
+     value: PropTypes.node,
+     classNameWrapper: PropTypes.string.isRequired,
+   }
+
+   static defaultProps = {
+     value: '',
+   }
+
+   render() {
+     const {
+       forID,
+       value,
+       onChange,
+       classNameWrapper,
+     } = this.props;
+
+     return (
+       <input
+         type="text"
+         id={forID}
+         className={classNameWrapper}
+         value={value || ''}
+         onChange={e => onChange(e.target.value)}
+       />
+     );
+   }
+ }
+```
+
+`src/Preview.js`
+
+```javascript
+import PropTypes from 'prop-types';
+import React from 'react';
+
+export default function Preview({ value }) {
+  return <div>{ value }</div>;
+}
+
+Preview.propTypes = {
+  value: PropTypes.node,
+};
+```
+
+`src/index.js`
+
+```javascript
+import Control from './Control'
+import Preview from './Preview'
+
+if (typeof window !== 'undefined') {
+  window.Control = Control
+  window.Preview = Preview
+}
+
+export { Control, Preview }
+```
+
+6. Change the content of `package.json` as below:
```

Comment:
Instead of copy pasting the entire file, I think we should split it into steps that explain the process.
For example, we should have a "setup wepback" after:
https://github.com/netlify/netlify-cms/pull/4146/files#diff-71cf0183902b4785a350655ed78e3fb2R215

that says, "In order to build React components we need to set up a build step. We'll be using Webpack. Please run the following command to install the required dependencies..."

Then another step to setting up the actual Webpack config, then we can explain the creation of the index file to export the control component and preview component.

To summarize, describe the process of generating the different files and the reason behind it.

Reply:
I have made changes. 
But I couldn't find how can I install "peerDependencies" from the command. ("peerDependencies": {},)
Should we add it manually to package.json?
I have searched tons of documents but [they say](https://stackoverflow.com/questions/51293822/webpack-cli3-0-8-requires-a-peer-of-webpack4-x-x-but-none-is-installed) if you install the webpack, peerDependencies will fill automatically.
But when I try to install all dependencies from scratch, it did not happen for me.
It should be like that:
"peerDependencies": {
    "react": "^16"
  },
Could you help me?

Reply:
Thanks @KancerEzeroglu. I appreciate the effort. We can add it manually.

Reply:
I've made changes, it is ready for review. fyi